test/batcher_test.go

```diff
@@ -525,3 +527,290 @@ func TestSecondaryBatcherRestartRecover(t *testing.T) {
+		require.NoError(t, err)
+
+		// Modify the batcher configuration to require client signature verification.
+		conf.NodeLocalConfig.GeneralConfig.ClientSignatureVerificationRequired = true
+		utils.WriteToYAML(conf.NodeLocalConfig, configFilePath)
+	}
+
+	// 5. Run the arma nodes.
+	// NOTE: if one of the nodes is not started within 10 seconds, there is no point in continuing the test, so fail it
+	// Obtains a test user configuration and constructs a broadcast client.
+	readyChan := make(chan struct{}, numOfArmaNodes)
+	armaNetwork := testutil.RunArmaNodes(t, dir, armaBinaryPath, readyChan, netInfo)
+	defer armaNetwork.Stop()
+
+	testutil.WaitReady(t, readyChan, numOfArmaNodes, 10)
+
+	uc, err := testutil.GetUserConfig(dir, types.PartyID(1))
+	assert.NoError(t, err)
+	assert.NotNil(t, uc)
+
+	totalTxNumber := 0
+	txNumber := 10
+	// rate limiter parameters
+	fillInterval := 10 * time.Millisecond
+	fillFrequency := 1000 / int(fillInterval.Milliseconds())
+	rate := 100
+
+	capacity := rate / fillFrequency
+	rl, err := armageddon.NewRateLimiter(rate, fillInterval, capacity)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to start a rate limiter")
+		os.Exit(3)
+	}
+
+	broadcastClient := client.NewBroadcastTxClient(uc, 10*time.Second)
+	signer, certBytes, err := testutil.LoadCryptoMaterialsFromDir(t, uc.MSPDir)
+	require.NoError(t, err)
+
+	for i := range txNumber {
+		status := rl.GetToken()
+		if !status {
+			fmt.Fprintf(os.Stderr, "failed to send tx %d", i+1)
+			os.Exit(3)
+		}
+		txContent := tx.PrepareTxWithTimestamp(totalTxNumber+i, 64, []byte("sessionNumber"))
+		env := tx.CreateSignedStructuredEnvelope(txContent, signer, certBytes, "org1")
+		err = broadcastClient.SendTx(env)
+		require.NoError(t, err)
+	}
+
+	totalTxNumber += txNumber
+
+	// 6. Make sure assemblers of all the parties get transactions (expect 10 TXs).
+
+	parties := []types.PartyID{}
+	for partyID := range numOfParties {
+		parties = append(parties, types.PartyID(partyID+1))
+	}
+
+	blockPullerInfo := PullFromAssemblers(t, &BlockPullerOptions{
+		UserConfig:       uc,
+		Parties:          parties,
+		StartBlock:       0,
+		EndBlock:         math.MaxUint64,
+		Transactions:     totalTxNumber,
+		Timeout:          60,
+		NeedVerification: true,
+		ErrString:        "cancelled pull from assembler: %d",
+	})
+
+	// 7. Determine a primary/non-primary party ID
+	var nonPrimaryPartyID, primaryPartyID types.PartyID
+
+	for partyID, bpi := range blockPullerInfo {
+		if bpi.Primary[types.ShardID(1)] != partyID {
+			nonPrimaryPartyID = partyID
+			break
+		}
+	}
+
+	require.NotZero(t, nonPrimaryPartyID, "Non-primary party ID should not be zero")
+
+	primaryPartyID = blockPullerInfo[nonPrimaryPartyID].Primary[types.ShardID(1)]
+
+	t.Logf("Non-primary party ID is %d", nonPrimaryPartyID)
+	t.Logf("Primary party ID is %d", primaryPartyID)
+
+	// 8. Send TXs to a non-primary batcher
+	stream := broadcastClient.SendStream(nonPrimaryPartyID)
+	require.NotNil(t, stream)
```

Comment:
Instead of exposing the underlying streams, lets simple add a method to `broadcastClient` that sends to a particular party rather than all:

`SendTxTo(envelope, partyID) error`

Internally it will lookup the respective stream and send on it.